src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -513,24 +349,55 @@ std::optional<DataModel::ActionReturnStatus> CodegenDataModelProvider::Invoke(co
     return std::nullopt;
 }
 
-bool CodegenDataModelProvider::EndpointExists(EndpointId endpoint)
+DataModel::MetadataList<DataModel::EndpointEntry> CodegenDataModelProvider::Endpoints()
 {
-    return (emberAfIndexFromEndpoint(endpoint) != kEmberInvalidEndpointIndex);
-}
+    DataModel::MetadataList<DataModel::EndpointEntry> result;
 
-std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfo(EndpointId endpoint)
-{
-    std::optional<unsigned> endpoint_idx = TryFindEndpointIndex(endpoint);
-    if (endpoint_idx.has_value())
+    const uint16_t endpointCount = emberAfEndpointCount();
+
+    // allocate the max as some endpoints may be disabled
+    CHIP_ERROR err = result.reserve(endpointCount);
+    if (err != CHIP_NO_ERROR)
     {
-        return GetEndpointInfoAtIndex(static_cast<uint16_t>(*endpoint_idx));
+#if CHIP_ERROR_LOGGING && CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
+        ChipLogError(AppServer, "Failed to allocate space for endpoints: %" CHIP_ERROR_FORMAT, err.Format());
+#endif
+        return {};
     }
-    return std::nullopt;
-}
 
-DataModel::EndpointEntry CodegenDataModelProvider::FirstEndpoint()
-{
-    return FirstEndpointEntry(0, mEndpointIterationHint);
+    for (uint16_t endpointIndex = 0; endpointIndex < endpointCount; endpointIndex++)
+    {
+        if (!emberAfEndpointIndexIsEnabled(endpointIndex))
+        {
+            continue;
+        }
+
+        DataModel::EndpointEntry entry;
+        entry.id       = emberAfEndpointFromIndex(endpointIndex);
+        entry.parentId = emberAfParentEndpointFromIndex(endpointIndex);
+
+        switch (GetCompositionForEndpointIndex(endpointIndex))
+        {
+        case EndpointComposition::kFullFamily:
+            entry.compositionPattern = DataModel::EndpointCompositionPattern::kFullFamily;
+            break;
+        case EndpointComposition::kTree:
+        default:
```

Comment:
If you don't have a default case, compiler should catch new enum values being added, right?

Reply:
Enumeration had a kInvalid entry. Added that to the list instead.